contracts/cw1-subkeys/src/contract.rs

```diff
@@ -239,11 +263,115 @@ mod tests {
 
     #[test]
     fn increase_allowances() {
-        // TODO
-        // add to existing account (expires = None) => don't change Expiration (previous should be different than Never)
-        // add to existing account (expires = Some)
-        // add to new account (expires = None) => default Expiration::Never
-        // add to new account (expires = Some)
+        let mut deps = mock_dependencies(20, &coins(1111, "token1"));
+
+        let owner = HumanAddr::from("admin0001");
+        let admins = vec![owner.clone(), HumanAddr::from("admin0002")];
+
+        let spender1 = HumanAddr::from("spender0001");
+        let spender2 = HumanAddr::from("spender0002");
+        let spender3 = HumanAddr::from("spender0003");
+        let spender4 = HumanAddr::from("spender0004");
+        let spenders = vec![spender1.clone(), spender2.clone()];
+
+        // Same allowances for all spenders, for simplicity
+        let denom1 = "token1";
+        let denom2 = "token2";
+        let amount1 = 1111;
+        let amount2 = 2222;
+        let allowances = vec![coin(amount1, denom1), coin(amount2, denom2)];
+
+        let expires_height = Expiration::AtHeight { height: 5432 };
+        let expires_never = Expiration::Never {};
+        let expires_time = Expiration::AtTime { time: 1234567890};
+        // Initially set first spender allowance with height expiration, the second with no expiration
+        let expirations = vec![expires_height.clone(), expires_never.clone()];
+
+        let allow1 = &allowances[0];
+        let allow2 = &allowances[1];
+
+        let env = mock_env(owner, &[]);
+        setup_test_case(
+            &mut deps,
+            &env,
+            &admins,
+            &spenders,
+            &allowances,
+            &expirations,
+        );
+
+        // Add to spender1 account (expires = None) => don't change Expiration
+        let msg = HandleMsg::IncreaseAllowance {
+            spender: spender1.clone(),
+            amount: allow1.clone(),
+            expires: None,
+        };
+        handle(&mut deps, env.clone(), msg).unwrap();
+
+        // Verify
+        let allowance = query_allowance(&deps, spender1.clone()).unwrap();
+        assert_eq!(
+            allowance,
+            Allowance {
+                balance: Balance(vec![coin(amount1 * 2, &allow1.denom), allow2.clone()]),
+                expires: expires_height.clone()
+            }
+        );
+
+        // Add to spender2 account (expires = Some)
+        let msg = HandleMsg::IncreaseAllowance {
+            spender: spender2.clone(),
+            amount: allow1.clone(),
+            expires: Some(expires_height.clone()),
+        };
+        handle(&mut deps, env.clone(), msg).unwrap();
+
+        // Verify
+        let allowance = query_allowance(&deps, spender2.clone()).unwrap();
+        assert_eq!(
+            allowance,
+            Allowance {
+                balance: Balance(vec![coin(amount1 * 2, &allow1.denom), allow2.clone()]),
```

Comment:
for the sake of completeness, could you add a different denom here, either allow2, or maybe even allow3?

Reply:
Sure, will do it in the continuation PR.